### PR TITLE
P2533R0 Core Language Working Group "ready"

### LINKED ISSUE #5257: [2022-02 CWG Motion 1] P2533R0 (Core Language Working Group "ready" Issue
[P2533R0](https://wg21.link/p2533r0)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5257 
Fixes #4843
Fixes #2338
Fixes cplusplus/papers#1196

### CODE REVIEW

File: source/classes.tex [hunk 1]
```diff
@@ -2473,7 +2473,6 @@
 \indextext{conversion!user-defined}%
 
 \pnum
```
Comment (tkoeppe) [inline source/classes.tex:2475]: This is an odd (ineffective?) `\pnum` before a pure grammar block?

**Review by tkoeppe — CHANGES_REQUESTED** [review]

File: source/classes.tex [hunk 2]
```diff
@@ -2496,7 +2496,7 @@
 its \grammarterm{declarator} shall be
 a function declarator\iref{dcl.fct} of the form
 \begin{ncsimplebnf}
-ptr-declarator \terminal{(} parameter-declaration-clause \terminal{)} \opt{cv-qualifier-seq}
+ptr-declarator \terminal{(} parameter-declaration-clause \terminal{)} \opt{cv-qualifier-seq}\br
```
Comment (tkoeppe) [inline source/classes.tex:2499]: Is this change in the wrong commit?

File: source/expressions.tex [hunk 3]
```diff
@@ -1690,7 +1690,20 @@
...
+\nontermdef{lambda-specifier}\br
+    \keyword{consteval}\br
+    \keyword{constexpr}\br
+    \keyword{mutable}
+\end{bnf}
+
+\begin{bnf}
+\nontermdef{lambda-specifier-seq}\br
+    lambda-specifier\br
+    lambda-specifier lambda-specifier-seq\br
```
Comment (tkoeppe) [inline source/expressions.tex:1706]: Why a traling `\br` here but not above?

**Review by tkoeppe — CHANGES_REQUESTED** [review]

File: source/expressions.tex (hunk 3 above)
Comment (jensmaurer) [reply source/expressions.tex:1706]: There should not be a trailing `\br` here.

If you want to, I can fix that and force-push.

File: source/classes.tex (hunk 1 above)
Comment (jensmaurer) [reply source/classes.tex:2475]: Yes, this should have been deleted.

If you want to, I can fix that and force-push.

File: source/classes.tex (hunk 2 above)
Comment (jensmaurer) [reply source/classes.tex:2499]: Yes, this should be squashed into "CWG1726 Declarator operators and conversion functions".

If you want to, I can fix that and force-push.

File: source/classes.tex (hunk 1 above)
Comment (jensmaurer) [reply source/classes.tex]: fixed

File: source/classes.tex (hunk 2 above)
Comment (jensmaurer) [reply source/classes.tex:2498]: fixed

File: source/expressions.tex (hunk 3 above)
Comment (jensmaurer) [reply source/expressions.tex]: fixed

**Review by tkoeppe — APPROVED** [review]

File: source/classes.tex [hunk 4]
```diff
@@ -2488,18 +2486,52 @@
...
+the \grammarterm{member-specification} of a class or class template
+but is not a friend declaration\iref{class.friend},
+the \grammarterm{id-expression} is a \grammarterm{conversion-function-id};
+\item
+otherwise, the \grammarterm{id-expression} is a \grammarterm{qualified-id}
+whose \grammarterm{unqualified-id} is a \grammarterm{conversion-function-id}.
+\end{itemize}
+
+\pnum
+A conversion function  shall have no non-object parameters and
```
Comment (JohelEGP) [inline source/classes.tex:2518]: The core issue doesn't include "non-object".

I suppose you added it to account for explicit object parameters.

Is that really enough? No need to restrict it to an explicit object parameter, but any amount of non-object parameters is OK? Is there wording elsewhere that make these cases ill-formed, and thus saying "non-object" is enough?

How about the function type specified below? Does it need a revision, too?

Comment (tkoeppe) [reply source/classes.tex:2518]: I think the paper misrepresents the status quo? So tihs seems like a "plausible merge".

Comment (jensmaurer) [reply source/classes.tex:2518]: A conversion function with an explicit this parameter seems plausible, but there must not be other parameters. Further cleanup (if any) should be in a separate editorial or core issue.

**Pull request merged**